memory/jemalloc_nodump_allocator.h

```diff
@@ -78,17 +96,304 @@ class JemallocNodumpAllocator : public BaseMemoryAllocator {
+#endif  // ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+
+static std::unordered_map<std::string, OptionTypeInfo> jemalloc_type_info = {
+    {"limit_tcache_size",
+     {offsetof(struct JemallocAllocatorOptions, limit_tcache_size),
+      OptionType::kBoolean, OptionVerificationType::kNormal,
+      OptionTypeFlags::kNone}},
+    {"tcache_size_lower_bound",
+     {offsetof(struct JemallocAllocatorOptions, tcache_size_lower_bound),
+      OptionType::kSizeT, OptionVerificationType::kNormal,
+      OptionTypeFlags::kNone}},
+    {"tcache_size_upper_bound",
+     {offsetof(struct JemallocAllocatorOptions, tcache_size_upper_bound),
+      OptionType::kSizeT, OptionVerificationType::kNormal,
+      OptionTypeFlags::kNone}},
+};
+
+template <size_t kLog2NumArenas>
+bool JemallocNodumpAllocator<kLog2NumArenas>::IsSupported(std::string* why) {
+#ifndef ROCKSDB_JEMALLOC
+  *why = "Not compiled with ROCKSDB_JEMALLOC";
+  return false;
+#else
+  static const std::string unsupported =
+      "JemallocNodumpAllocator only available with jemalloc version >= 5 "
+      "and MADV_DONTDUMP is available.";
+  if (!HasJemalloc()) {
+    *why = unsupported;
+    return false;
+  }
+#ifndef ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+  *why = unsupported;
+  return false;
+#else
+  return true;
+#endif  // ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+#endif  // ROCKSDB_MALLOC
+}
+
+template <size_t kLog2NumArenas>
+JemallocNodumpAllocator<kLog2NumArenas>::JemallocNodumpAllocator(
+    JemallocAllocatorOptions& options)
+    : options_(options)
+#ifdef ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+      ,
+      tcache_(&JemallocNodumpAllocator<
+              kLog2NumArenas>::DestroyThreadSpecificCache) {
+#else   // ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+{
+#endif  // ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+  RegisterOptions(&options_, &jemalloc_type_info);
+}
+
+#ifdef ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+template <size_t kLog2NumArenas>
+JemallocNodumpAllocator<kLog2NumArenas>::~JemallocNodumpAllocator() {
+  // Destroy tcache before destroying arena.
+  autovector<void*> tcache_list;
+  tcache_.Scrape(&tcache_list, nullptr);
+  for (void* tcache_index : tcache_list) {
+    DestroyThreadSpecificCache(tcache_index);
+  }
+  // Destroy created arenas (assumed to have nonzero indexes). Silently ignore
+  // errors.
+  for (auto arena_index : arena_indexes_) {
+    if (arena_index == 0) {
+      continue;
+    }
+    assert(init_);
+    Status s = DestroyArena(arena_index);
+    assert(s.ok());
+    s.PermitUncheckedError();
+  }
+}
+
+template <size_t kLog2NumArenas>
+size_t JemallocNodumpAllocator<kLog2NumArenas>::UsableSize(
+    void* p, size_t /*allocation_size*/) const {
+  return malloc_usable_size(static_cast<void*>(p));
+}
+
+template <size_t kLog2NumArenas>
+void* JemallocNodumpAllocator<kLog2NumArenas>::Allocate(size_t size) {
+  // `size` is used as a source of entropy to initialize each thread's arena
+  // selection.
+  thread_local uint32_t tl_rng_seed = static_cast<uint32_t>(size);
```

Comment:
If it's an open question whether core local would be more efficient here, maybe drop a comment saying it's a possible investigation item.